### ISSUE TITLE
Support Location type that does not implement Default

### ISSUE THREAD

**Issue by dtolnay** [nikomatsakis/lalrpop#285]
I am working on parsing a TokenStream in a procedural macro using LALRPOP with [`proc_macro::Span`](https://doc.rust-lang.org/proc_macro/struct.Span.html) as my Location, but there is no Default impl. Would it be possible to provide an alternative way for LALRPOP to initialize a default value? In a procedural macro I expect this would be `Span::call_site()`.

Currently the generated code seems to invoke Default in two places.

```
error[E0277]: the trait bound `Span: std::default::Default` is not satisfied
  --> src/tokens.rs:93:35
   |
93 |         let mut __last_location = Default::default();
   |                                   ^^^^^^^^^^^^^^^^ the trait `std::default::Default` is not implemented for `Span`
   |
   = note: required by `std::default::Default::default`

error[E0599]: no method named `unwrap_or_default` found for type `std::option::Option<Span>` in the current scope
   --> src/tokens.rs:209:69
    |
209 |                 let __start = __symbols.last().map(|s| s.2.clone()).unwrap_or_default();
    |                                                                     ^^^^^^^^^^^^^^^^^
    |
    = note: the method `unwrap_or_default` exists but the following trait bounds were not satisfied:
            `Span : std::default::Default`
```

**Comment by dtolnay** [comment]
I ended up having to [wrap my location type in a newtype that provides `Default`](https://github.com/dtolnay/lalrproc/blob/d3b9148f8c42cfc8b72d65a2f76e8ab4cc8e80f5/src/span.rs) to work around this.

**Comment by nikomatsakis** [comment]
Yeah, we could provide some alternative way. Maybe I should add a trait to `lalrpop_util` with a blanket impl for things that implement `Default` -- it's basically a way to get the location indicating "start of the stream".

**Comment by nikomatsakis** [comment]
Alternatively, I could ask you to provide that value explicitly I suppose, but that'd be kind of annoying to thread around.

**Comment by nikomatsakis** [comment]
So locations wind up having to implement `Default` because sometimes we have to synthesize a location to represent the "start of the stream". This comes from a few spots in the code:

https://github.com/lalrpop/lalrpop/blob/5246910e7e656c56f0b77c4f8c3cf33862b1c493/lalrpop/src/lr1/codegen/parse_table.rs#L839-L842

https://github.com/lalrpop/lalrpop/blob/5246910e7e656c56f0b77c4f8c3cf33862b1c493/lalrpop/src/lr1/codegen/parse_table.rs#L484-L486

https://github.com/lalrpop/lalrpop/blob/5246910e7e656c56f0b77c4f8c3cf33862b1c493/lalrpop/src/lr1/codegen/parse_table.rs#L1394

https://github.com/lalrpop/lalrpop/blob/5246910e7e656c56f0b77c4f8c3cf33862b1c493/lalrpop/src/lr1/codegen/ascent.rs#L785-L789

https://github.com/lalrpop/lalrpop/blob/5246910e7e656c56f0b77c4f8c3cf33862b1c493/lalrpop/src/lr1/codegen/ascent.rs#L760

(The last one is just a comment that would need to be updated.)

I think we should to introduce into `lalrpop_util` a trait like

```rust
/// Trait that types which want to be used as locations must implement.
trait Location { 
    fn start_location() -> Self;
}

impl<T: Default> Location for T {
    fn start_location() -> Self {
        Self::default()
    }
}
```

Hmm, but thinking on it a bit more, I'm not 100% sure this is the best solution, though, because -- in the absence of specialization -- these kinds of blanket impls can be annoying, and may well force people like @dtolnay to introduce newtypes themselves. 

An alternative -- which is harder -- would be to modify the generated code to take a "starting position" argument and to thread that value down and clone it where needed. Not so sure about this either.

**Comment by dtolnay** [comment]
I agree that a trait with a default impl would be just as unworkable for some (slightly smaller) fraction of use cases and would not be the best fix.

A would prefer to take the approach of having an $expr that we paste instead of `Default::default()` into the generated code, rather than threading around the start location at runtime and cloning it. Conceptually the following, but it would need some design work because the mixing of items and statements in an `extern` block does not look great.

```rust
extern {
    type Location = Span;
    let start_location = Span::call_site();

    enum Token {
        /* ... */
    }
}
```

**Comment by ahmedcharles** [comment]
What's the downside of not having the blanket impl? Just having the trait should be fine, since the type can impl the trait. The only scenario where this wouldn't work is when it runs afoul of the orphan rule.

**Comment by ahmedcharles** [comment]
So, the options are:

1. Implement the Location using a new type and implement `Default`.
2. Allow an arbitrary expression.

Thanks for the explanation.

**Comment by ISSOtm** [comment]
In my use case, the default location depends on one of the parameters passed to the grammar; implementing a generic `Default` (or another trait) would require relying on global state, which I can currently accommodate but may become problematic in the future.
